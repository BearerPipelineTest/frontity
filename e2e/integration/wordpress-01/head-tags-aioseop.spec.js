describe("Head Tags - All in One SEO Pack", () => {
  before(() => {
    cy.task("installPlugin", { name: "all-in-one-seo-pack", version: "3.7.1" });
    cy.task("installPlugin", { name: "rest-api-head-tags" });
    cy.task("installPlugin", { name: "custom-post-type-ui" });
    cy.task("loadDatabase", {
      path: "./wp-data/head-tags/aioseop.sql",
    });
  });

  after(() => {
    cy.task("resetDatabase");
    cy.task("removeAllPlugins");
  });

  /**
   * Check the title for the current page is the given one.
   *
   * @param title - The page title for the given link.
   */
  const checkTitle = (title) => {
    it("should render the correct title", () => {
      cy.get("title").should("contain", title);
    });
  };

  /**
   * Ensure that the canonical link has been rendered.
   */
  const checkCanonical = () => {
    it("should render the correct canonical link", () => {
      cy.get('link[rel="canonical"]').toMatchSnapshot();
    });
  };

  /**
   * Ensure that the ld+json schema has been rendered.
   */
  const checkSchema = () => {
    it("should render the schema tag", () => {
      cy.get('script[type="application/ld+json"]').toMatchSnapshot();
    });
  };

  /**
   * Visit the specified link with scripts disabled.
   *
   * @param link - The link of the page.
   * @param options - Visit options.
   */
  const visitLink = (link, options) => {
    cy.visit(`http://localhost:3001${link}?frontity_name=head-tags`, options);
  };

  /**
   * Change the router value in Frontity.
   *
   * @param link - The link of the page.
   */
  const routerSet = (link) => {
    cy.window().then((win) => {
      win.frontity.actions.router.set(link);
    });
  };

  /**
   * Tests for posts.
   */
  describe("Post", () => {
    const link = "/hello-world/";
    const title = "Hello world! | Test WP Site";

    before(() => visitLink(link, { script: false }));
    checkTitle(title);
    checkCanonical();

    /**
     * The `checkSchema()` call was replaced here by a check that the schema
     * tag simply exists and it is not empty.
     *
     * The reason for this is that the tag content is not the same for different
     * versions of WordPress:
     * - 5.4 : `"commentsCount": "1",`.
     * - 5.5 : `"commentsCount": 1,`.
     */
    it("should render the schema tag", () => {
      cy.get('script[type="application/ld+json"]').should("not.be.empty");
    });
  });

  /**
   * Tests for pages.
   */
  describe("Page", () => {
    const link = "/sample-page/";
    const title = "Sample Page | Test WP Site";

    before(() => visitLink(link, { script: false }));
    checkTitle(title);
    checkCanonical();
    checkSchema();
  });

  /**
   * Tests for categories.
   */
  describe("Category", () => {
    const link = "/category/nature/";
    const title = "Nature | Test WP Site";

    before(() => visitLink(link, { script: false }));
    checkTitle(title);
    checkCanonical();
    checkSchema();
  });

  /**
   * Tests for tags.
   */
  describe("Tag", () => {
    const link = "/tag/japan/";
    const title = "Japan | Test WP Site";

    before(() => visitLink(link, { script: false }));
    checkTitle(title);
    checkCanonical();
    checkSchema();
  });

  /**
   * Tests for authors.
   */
  describe("Author", () => {
    const link = "/author/luisherranz";
    const title = "luisherranz | Test WP Site";

    before(() => visitLink(link, { script: false }));

    /**
     * We don't check the schema here because it is not included by the
     * _REST API - Head Tags_ plugin when used along with _All in One SEO Pack_,
     * because it is not generated correctly for authors.
     */
    checkTitle(title);
    checkCanonical();
  });

  /**
   * Tests for the homepage.
   */
  describe("Homepage", () => {
    const link = "/";
    const title = "Test WP Site | Just another WordPress site";

    before(() => visitLink(link, { script: false }));
    checkTitle(title);
    checkCanonical();
    checkSchema();
  });

  /**
   * Tests for custom post types.
   */
  describe("CPT", () => {
    const link = "/movie/the-terminator/";
    const title = "The Terminator | Test WP Site";

    before(() => visitLink(link, { script: false }));
    checkTitle(title);
    checkCanonical();
    checkSchema();
  });

  /**
   * Tests for archive of custom post types.
   */
  describe("CPT (archive)", () => {
    const title = "Movies | Test WP Site";
    const link = "/movies/";

    before(() => visitLink(link, { script: false }));
    checkTitle(title);
    checkCanonical();
    checkSchema();
  });

  /**
   * Tests for archive of custom post types.
   */
  describe("Custom Taxonomy", () => {
    const link = "/actor/linda-hamilton/";
    const title = "Linda Hamilton | Test WP Site";

    before(() => visitLink(link, { script: false }));
    checkTitle(title);
    checkCanonical();
    checkSchema();
  });

  /**
   * Test for the `<title>` tag while navigating through Frontity.
   */
  describe("Title tag", () => {
    it("should be correct while navigating", () => {
      visitLink("/");
      cy.get("title").should(
        "contain",
        "Test WP Site | Just another WordPress site"
      );

      routerSet("/hello-world/");
      cy.get("title").should("contain", "Hello world! | Test WP Site");

      routerSet("/sample-page/");
      cy.get("title").should("contain", "Sample Page | Test WP Site");

      routerSet("/category/nature/");
      cy.get("title").should("contain", "Nature | Test WP Site");

      routerSet("/tag/japan/");
      cy.get("title").should("contain", "Japan | Test WP Site");

      routerSet("/author/luisherranz");
      cy.get("title").should("contain", "luisherranz | Test WP Site");

      routerSet("/movie/the-terminator/");
      cy.get("title").should("contain", "The Terminator | Test WP Site");

      routerSet("/movies/");
      cy.get("title").should("contain", "Movies | Test WP Site");

      routerSet("/actor/linda-hamilton/");
      cy.get("title").should("contain", "Linda Hamilton | Test WP Site");
    });
  });
});
