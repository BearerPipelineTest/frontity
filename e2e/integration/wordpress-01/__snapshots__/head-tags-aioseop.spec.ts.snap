exports[`Head Tags - All in One SEO Pack > Post > should render the correct canonical link #0`] = `
<link
  data-rh="true"
  rel="canonical"
  href="http://my.frontity.site/hello-world/"
/>
`;

exports[`Head Tags - All in One SEO Pack > Post > should render the schema tag #0`] = `
<script data-rh="true" type="application/ld+json" class="aioseop-schema">
  {
    "@context": "https://schema.org",
    "@graph": [
      {
        "@type": "Organization",
        "@id": "http://my.frontity.site/#organization",
        "url": "http://my.frontity.site/",
        "name": "Test WP Site",
        "sameAs": []
      },
      {
        "@type": "WebSite",
        "@id": "http://my.frontity.site/#website",
        "url": "http://my.frontity.site/",
        "name": "Test WP Site",
        "publisher": { "@id": "http://my.frontity.site/#organization" },
        "potentialAction": {
          "@type": "SearchAction",
          "target": "http://my.frontity.site/?s={search_term_string}",
          "query-input": "required name=search_term_string"
        }
      },
      {
        "@type": "WebPage",
        "@id": "http://my.frontity.site/hello-world/#webpage",
        "url": "http://my.frontity.site/hello-world/",
        "inLanguage": "en-US",
        "name": "Hello world!",
        "isPartOf": { "@id": "http://my.frontity.site/#website" },
        "breadcrumb": {
          "@id": "http://my.frontity.site/hello-world/#breadcrumblist"
        },
        "datePublished": "2020-08-20T01:31:05+00:00",
        "dateModified": "2020-08-31T13:42:07+00:00"
      },
      {
        "@type": "Article",
        "@id": "http://my.frontity.site/hello-world/#article",
        "isPartOf": { "@id": "http://my.frontity.site/hello-world/#webpage" },
        "author": { "@id": "http://my.frontity.site/author/admin/#author" },
        "headline": "Hello world!",
        "datePublished": "2020-08-20T01:31:05+00:00",
        "dateModified": "2020-08-31T13:42:07+00:00",
        "commentCount": 1,
        "mainEntityOfPage": {
          "@id": "http://my.frontity.site/hello-world/#webpage"
        },
        "publisher": { "@id": "http://my.frontity.site/#organization" },
        "articleSection": "Uncategorized"
      },
      {
        "@type": "Person",
        "@id": "http://my.frontity.site/author/admin/#author",
        "name": "admin",
        "sameAs": [],
        "image": {
          "@type": "ImageObject",
          "@id": "http://my.frontity.site/#personlogo",
          "url": "http://2.gravatar.com/avatar/55502f40dc8b7c769880b10874abc9d0?s=96&d=mm&r=g",
          "width": 96,
          "height": 96,
          "caption": "admin"
        }
      },
      {
        "@type": "BreadcrumbList",
        "@id": "http://my.frontity.site/hello-world/#breadcrumblist",
        "itemListElement": [
          {
            "@type": "ListItem",
            "position": 1,
            "item": {
              "@type": "WebPage",
              "@id": "http://my.frontity.site/",
              "url": "http://my.frontity.site/",
              "name": "Test WP Site"
            }
          },
          {
            "@type": "ListItem",
            "position": 2,
            "item": {
              "@type": "WebPage",
              "@id": "http://my.frontity.site/hello-world/",
              "url": "http://my.frontity.site/hello-world/",
              "name": "Hello world!"
            }
          }
        ]
      }
    ]
  }
</script>
`;

exports[`Head Tags - All in One SEO Pack > Page > should render the correct canonical link #0`] = `
<link
  data-rh="true"
  rel="canonical"
  href="http://my.frontity.site/sample-page/"
/>
`;

exports[`Head Tags - All in One SEO Pack > Page > should render the schema tag #0`] = `
<script data-rh="true" type="application/ld+json" class="aioseop-schema">
  {
    "@context": "https://schema.org",
    "@graph": [
      {
        "@type": "Organization",
        "@id": "http://my.frontity.site/#organization",
        "url": "http://my.frontity.site/",
        "name": "Test WP Site",
        "sameAs": []
      },
      {
        "@type": "WebSite",
        "@id": "http://my.frontity.site/#website",
        "url": "http://my.frontity.site/",
        "name": "Test WP Site",
        "publisher": { "@id": "http://my.frontity.site/#organization" },
        "potentialAction": {
          "@type": "SearchAction",
          "target": "http://my.frontity.site/?s={search_term_string}",
          "query-input": "required name=search_term_string"
        }
      },
      {
        "@type": "WebPage",
        "@id": "http://my.frontity.site/sample-page/#webpage",
        "url": "http://my.frontity.site/sample-page/",
        "inLanguage": "en-US",
        "name": "Sample Page",
        "isPartOf": { "@id": "http://my.frontity.site/#website" },
        "breadcrumb": {
          "@id": "http://my.frontity.site/sample-page/#breadcrumblist"
        },
        "datePublished": "2020-08-20T01:31:05+00:00",
        "dateModified": "2020-08-20T01:31:05+00:00"
      },
      {
        "@type": "BreadcrumbList",
        "@id": "http://my.frontity.site/sample-page/#breadcrumblist",
        "itemListElement": [
          {
            "@type": "ListItem",
            "position": 1,
            "item": {
              "@type": "WebPage",
              "@id": "http://my.frontity.site/",
              "url": "http://my.frontity.site/",
              "name": "Test WP Site"
            }
          },
          {
            "@type": "ListItem",
            "position": 2,
            "item": {
              "@type": "WebPage",
              "@id": "http://my.frontity.site/sample-page/",
              "url": "http://my.frontity.site/sample-page/",
              "name": "Sample Page"
            }
          }
        ]
      }
    ]
  }
</script>
`;

exports[`Head Tags - All in One SEO Pack > Category > should render the correct canonical link #0`] = `
<link
  data-rh="true"
  rel="canonical"
  href="http://my.frontity.site/category/nature/"
/>
`;

exports[`Head Tags - All in One SEO Pack > Category > should render the schema tag #0`] = `
<script data-rh="true" type="application/ld+json" class="aioseop-schema">
  {
    "@context": "https://schema.org",
    "@graph": [
      {
        "@type": "Organization",
        "@id": "http://my.frontity.site/#organization",
        "url": "http://my.frontity.site/",
        "name": "Test WP Site",
        "sameAs": []
      },
      {
        "@type": "WebSite",
        "@id": "http://my.frontity.site/#website",
        "url": "http://my.frontity.site/",
        "name": "Test WP Site",
        "publisher": { "@id": "http://my.frontity.site/#organization" },
        "potentialAction": {
          "@type": "SearchAction",
          "target": "http://my.frontity.site/?s={search_term_string}",
          "query-input": "required name=search_term_string"
        }
      },
      {
        "@type": "CollectionPage",
        "@id": "http://my.frontity.site/category/nature/#collectionpage",
        "url": "http://my.frontity.site/category/nature/",
        "inLanguage": "en-US",
        "name": "Nature",
        "isPartOf": { "@id": "http://my.frontity.site/#website" },
        "breadcrumb": {
          "@id": "http://my.frontity.site/category/nature/#breadcrumblist"
        }
      },
      {
        "@type": "BreadcrumbList",
        "@id": "http://my.frontity.site/category/nature/#breadcrumblist",
        "itemListElement": [
          {
            "@type": "ListItem",
            "position": 1,
            "item": {
              "@type": "WebPage",
              "@id": "http://my.frontity.site/",
              "url": "http://my.frontity.site/",
              "name": "Test WP Site"
            }
          },
          {
            "@type": "ListItem",
            "position": 2,
            "item": {
              "@type": "WebPage",
              "@id": "http://my.frontity.site/category/nature/",
              "url": "http://my.frontity.site/category/nature/",
              "name": "Nature"
            }
          }
        ]
      }
    ]
  }
</script>
`;

exports[`Head Tags - All in One SEO Pack > Tag > should render the correct canonical link #0`] = `
<link
  data-rh="true"
  rel="canonical"
  href="http://my.frontity.site/tag/japan/"
/>
`;

exports[`Head Tags - All in One SEO Pack > Tag > should render the schema tag #0`] = `
<script data-rh="true" type="application/ld+json" class="aioseop-schema">
  {
    "@context": "https://schema.org",
    "@graph": [
      {
        "@type": "Organization",
        "@id": "http://my.frontity.site/#organization",
        "url": "http://my.frontity.site/",
        "name": "Test WP Site",
        "sameAs": []
      },
      {
        "@type": "WebSite",
        "@id": "http://my.frontity.site/#website",
        "url": "http://my.frontity.site/",
        "name": "Test WP Site",
        "publisher": { "@id": "http://my.frontity.site/#organization" },
        "potentialAction": {
          "@type": "SearchAction",
          "target": "http://my.frontity.site/?s={search_term_string}",
          "query-input": "required name=search_term_string"
        }
      },
      {
        "@type": "CollectionPage",
        "@id": "http://my.frontity.site/tag/japan/#collectionpage",
        "url": "http://my.frontity.site/tag/japan/",
        "inLanguage": "en-US",
        "name": "Japan",
        "isPartOf": { "@id": "http://my.frontity.site/#website" },
        "breadcrumb": {
          "@id": "http://my.frontity.site/tag/japan/#breadcrumblist"
        }
      },
      {
        "@type": "BreadcrumbList",
        "@id": "http://my.frontity.site/tag/japan/#breadcrumblist",
        "itemListElement": [
          {
            "@type": "ListItem",
            "position": 1,
            "item": {
              "@type": "WebPage",
              "@id": "http://my.frontity.site/",
              "url": "http://my.frontity.site/",
              "name": "Test WP Site"
            }
          },
          {
            "@type": "ListItem",
            "position": 2,
            "item": {
              "@type": "WebPage",
              "@id": "http://my.frontity.site/tag/japan/",
              "url": "http://my.frontity.site/tag/japan/",
              "name": "Japan"
            }
          }
        ]
      }
    ]
  }
</script>
`;

exports[`Head Tags - All in One SEO Pack > Author > should render the correct canonical link #0`] = `
<link
  data-rh="true"
  rel="canonical"
  href="http://my.frontity.site/author/luisherranz/"
/>
`;

exports[`Head Tags - All in One SEO Pack > Homepage > should render the correct canonical link #0`] = `
<link data-rh="true" rel="canonical" href="http://my.frontity.site/" />
`;

exports[`Head Tags - All in One SEO Pack > Homepage > should render the schema tag #0`] = `
<script data-rh="true" type="application/ld+json" class="aioseop-schema">
  {
    "@context": "https://schema.org",
    "@graph": [
      {
        "@type": "Organization",
        "@id": "http://my.frontity.site/#organization",
        "url": "http://my.frontity.site/",
        "name": "Test WP Site",
        "sameAs": []
      },
      {
        "@type": "WebSite",
        "@id": "http://my.frontity.site/#website",
        "url": "http://my.frontity.site/",
        "name": "Test WP Site",
        "publisher": { "@id": "http://my.frontity.site/#organization" },
        "potentialAction": {
          "@type": "SearchAction",
          "target": "http://my.frontity.site/?s={search_term_string}",
          "query-input": "required name=search_term_string"
        }
      },
      {
        "@type": "WebPage",
        "@id": "http://my.frontity.site/#webpage",
        "url": "http://my.frontity.site/",
        "inLanguage": "en-US",
        "name": "Test WP Site",
        "isPartOf": { "@id": "http://my.frontity.site/#website" },
        "breadcrumb": { "@id": "http://my.frontity.site/#breadcrumblist" },
        "description": "Just another WordPress site",
        "about": { "@id": "http://my.frontity.site/#organization" }
      },
      {
        "@type": "BreadcrumbList",
        "@id": "http://my.frontity.site/#breadcrumblist",
        "itemListElement": [
          {
            "@type": "ListItem",
            "position": 1,
            "item": {
              "@type": "WebPage",
              "@id": "http://my.frontity.site/",
              "url": "http://my.frontity.site/",
              "name": "Test WP Site"
            }
          }
        ]
      }
    ]
  }
</script>
`;

exports[`Head Tags - All in One SEO Pack > CPT > should render the correct canonical link #0`] = `
<link
  data-rh="true"
  rel="canonical"
  href="http://my.frontity.site/movie/the-terminator/"
/>
`;

exports[`Head Tags - All in One SEO Pack > CPT > should render the schema tag #0`] = `
<script data-rh="true" type="application/ld+json" class="aioseop-schema">
  {}
</script>
`;

exports[`Head Tags - All in One SEO Pack > CPT (archive) > should render the correct canonical link #0`] = `
<link data-rh="true" rel="canonical" href="http://my.frontity.site/movies/" />
`;

exports[`Head Tags - All in One SEO Pack > CPT (archive) > should render the schema tag #0`] = `
<script data-rh="true" type="application/ld+json" class="aioseop-schema">
  {
    "@context": "https://schema.org",
    "@graph": [
      {
        "@type": "Organization",
        "@id": "http://my.frontity.site/#organization",
        "url": "http://my.frontity.site/",
        "name": "Test WP Site",
        "sameAs": []
      },
      {
        "@type": "WebSite",
        "@id": "http://my.frontity.site/#website",
        "url": "http://my.frontity.site/",
        "name": "Test WP Site",
        "publisher": { "@id": "http://my.frontity.site/#organization" },
        "potentialAction": {
          "@type": "SearchAction",
          "target": "http://my.frontity.site/?s={search_term_string}",
          "query-input": "required name=search_term_string"
        }
      },
      {
        "@type": "CollectionPage",
        "@id": "http://my.frontity.site/movies/#collectionpage",
        "url": "http://my.frontity.site/movies/",
        "inLanguage": "en-US",
        "name": "Movies",
        "isPartOf": { "@id": "http://my.frontity.site/#website" },
        "breadcrumb": {
          "@id": "http://my.frontity.site/movies/#breadcrumblist"
        }
      },
      {
        "@type": "BreadcrumbList",
        "@id": "http://my.frontity.site/movies/#breadcrumblist",
        "itemListElement": [
          {
            "@type": "ListItem",
            "position": 1,
            "item": {
              "@type": "WebPage",
              "@id": "http://my.frontity.site/",
              "url": "http://my.frontity.site/",
              "name": "Test WP Site"
            }
          },
          {
            "@type": "ListItem",
            "position": 2,
            "item": {
              "@type": "WebPage",
              "@id": "http://my.frontity.site/movies/",
              "url": "http://my.frontity.site/movies/",
              "name": "Movies"
            }
          }
        ]
      }
    ]
  }
</script>
`;

exports[`Head Tags - All in One SEO Pack > Custom Taxonomy > should render the correct canonical link #0`] = `
<link
  data-rh="true"
  rel="canonical"
  href="http://my.frontity.site/actor/linda-hamilton/"
/>
`;

exports[`Head Tags - All in One SEO Pack > Custom Taxonomy > should render the schema tag #0`] = `
<script data-rh="true" type="application/ld+json" class="aioseop-schema">
  {}
</script>
`;
